File: whiteboard/app/src/main/java/com/ditto/whiteboard/transport/DiagnosticsState.kt
package com.ditto.whiteboard.transport

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * Owns the transport's [TransportDiagnostics] snapshot and guarantees every mutation is an atomic
 * compare-and-set via [MutableStateFlow.update]. The transport touches diagnostics from many
 * coroutines at once (reliable/live/rate/presence workers on [kotlinx.coroutines.Dispatchers.Default]
 * plus connect/outbox/bind callbacks on the IO pool); a plain `value = value.copy(...)`
 * read-modify-write would let overlapping writers clobber each other and silently drop peer entries,
 * rates, snapshot progress or error text. Funnelling every write through here keeps the same
 * lost-update discipline `BoardSession` already uses for its board/preview state.
 *
 * Extracted so the diagnostics bookkeeping can be unit-tested in pure JVM without standing up a real
 * Ditto endpoint.
 */
internal class DiagnosticsState(initial: TransportDiagnostics) {
  private val state = MutableStateFlow(initial)
  val flow: StateFlow<TransportDiagnostics> = state.asStateFlow()
  val value: TransportDiagnostics get() = state.value

  /** Atomically mutate the whole diagnostics snapshot. */
  fun update(transform: (TransportDiagnostics) -> TransportDiagnostics) = state.update(transform)

  /** Atomically create-or-update a single peer entry. */
  fun updatePeer(peer: String, update: (PeerDiagnostics) -> PeerDiagnostics) {
    state.update { current ->
      val peerValue = update(current.peers[peer] ?: PeerDiagnostics(peerKey = peer))
      current.copy(peers = current.peers + (peer to peerValue))
    }
  }

  /**
   * Drop every peer entry whose key is not in [visiblePeers]. Called when presence changes so a peer
   * that left the mesh cannot linger in the connected-people count or presence graph.
   */
  fun retainPeers(visiblePeers: Set<String>) {
    state.update { current ->
      if (current.peers.keys.all { it in visiblePeers }) current
      else current.copy(peers = current.peers.filterKeys { it in visiblePeers })
    }
  }
}

File: whiteboard-apple/Packages/WhiteboardCore/Sources/WhiteboardCore/Transport/DiagnosticsState.swift
import Foundation

/// Owns the transport's `TransportDiagnostics` snapshot and guarantees every mutation is atomic.
/// The transport touches diagnostics from many tasks at once (reliable/live/rate/presence workers
/// plus connect/outbox/bind callbacks); a plain read-modify-write would let overlapping writers
/// clobber each other and silently drop peer entries, rates, snapshot progress or error text.
/// Actor isolation keeps the same lost-update discipline `BoardSession` uses for its board state.
public actor DiagnosticsState {
  public private(set) var value: TransportDiagnostics
  private var continuations: [UUID: AsyncStream<TransportDiagnostics>.Continuation] = [:]

  public init(initial: TransportDiagnostics) {
    value = initial
  }

  /// StateFlow-like stream: replays the current snapshot, then subsequent changes.
  public var updates: AsyncStream<TransportDiagnostics> {
    let (stream, continuation) = AsyncStream<TransportDiagnostics>.makeStream()
    let id = UUID()
    continuations[id] = continuation
    continuation.onTermination = { [weak self] _ in
      Task { await self?.removeContinuation(id) }
    }
    continuation.yield(value)
    return stream
  }

  /// Atomically mutate the whole diagnostics snapshot.
  public func update(_ transform: (TransportDiagnostics) -> TransportDiagnostics) {
    let next = transform(value)
    guard next != value else { return }
    value = next
    for continuation in continuations.values { continuation.yield(next) }
  }

  /// Atomically create-or-update a single peer entry.
  public func updatePeer(_ peer: String, _ update: (PeerDiagnostics) -> PeerDiagnostics) {
    self.update { current in
      var next = current
      next.peers[peer] = update(current.peers[peer] ?? PeerDiagnostics(peerKey: peer))
      return next
    }
  }

  /// Drop every peer entry whose key is not in `visiblePeers`. Called when presence changes so a
  /// peer that left the mesh cannot linger in the connected-people count or presence graph.
  public func retainPeers(_ visiblePeers: Set<String>) {
    self.update { current in
      guard !current.peers.keys.allSatisfy({ visiblePeers.contains($0) }) else { return current }
      var next = current
      next.peers = next.peers.filter { visiblePeers.contains($0.key) }
      return next
    }
  }

  private func removeContinuation(_ id: UUID) {
    continuations[id] = nil
  }
}
